unittests/Dialect/AIG/LongestPathTest.cpp

```diff
@@ -0,0 +1,154 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#include "circt/Dialect/AIG/AIGDialect.h"
+#include "circt/Dialect/AIG/AIGOps.h"
+#include "circt/Dialect/Comb/CombDialect.h"
+#include "circt/Dialect/HW/HWDialect.h"
+#include "circt/Dialect/Seq/SeqDialect.h"
+
+#include "circt/Dialect/AIG/Analysis/LongestPathAnalysis.h"
+#include "circt/Dialect/Seq/SeqOps.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Pass/AnalysisManager.h"
+#include "llvm/ADT/SmallVector.h"
+#include "gtest/gtest.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace aig;
+
+namespace {
+
+const char *ir = R"MLIR(
+    hw.module private @basic(in %clock : !seq.clock, in %a : i2, in %b : i2, out x : i2, out y: i4) {
+      %p = seq.firreg %a clock %clock : i2
+      %q = seq.firreg %s clock %clock : i2
+      %r = hw.instance "inst" @child(a: %p: i2, b: %b: i2) -> (x: i2)
+      %s = aig.and_inv not %p, %q, %r : i2
+      %dummy = comb.concat %s, %a : i2, i2
+      hw.output %s, %dummy : i2, i4
+    } 
+    hw.module private @child(in %a : i2, in %b : i2, out x : i2) {
+      %r = aig.and_inv not %a, %b {sv.namehint = "child.r"} : i2
+      hw.output %r : i2
+    }
+    )MLIR";
```

Comment:
Do you know if there is some precedence for having MLIR blobs encoded in the C++ source as strings? This feels sensible since you are trying to go through a lot of the details provided by the analysis. On the other hand, you could argue that maybe the printer should have a verbose option where it prints all the interesting info you'd want to check, which would then allow you to check this with FileCheck.

Reply:
MLIR unittest use this pattern so I think it's ok to use it. 
https://github.com/llvm/llvm-project/blob/c6c2e21028cadef854cf22f6ecaa5eb9d224b76d/mlir/unittests/ExecutionEngine/Invoke.cpp#L74

> On the other hand, you could argue that maybe the printer should have a verbose option where it prints all the interesting info you'd want to check, which would then allow you to check this with FileCheck.

That's a good point but I think there are several things easier to test with C++ unittests, e.g. analysis has a debug point flag (analysis printer pass uses a debug flag because it's necessary to print traces). Also for retiming pass I have a plan to add an API to override visitors from analysis constructor so it will be necessary to write a test for that as well.